app/feature/feature-insurances/src/main/kotlin/com/hedvig/android/feature/insurances/terminatedcontracts/TerminatedContractsViewModel.kt

```diff
@@ -1,75 +1,87 @@
-import kotlinx.coroutines.flow.StateFlow
-import kotlinx.coroutines.flow.WhileSubscribed
-import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.collectLatest
 import kotlinx.coroutines.flow.map
 import kotlinx.coroutines.flow.onEach
-import kotlinx.coroutines.flow.stateIn
 
 internal class TerminatedContractsViewModel(
+  getInsuranceContractsUseCaseProvider: Provider<GetInsuranceContractsUseCase>,
+) : MoleculeViewModel<TerminatedContractsEvent, TerminatedContractsUiState>(
+    initialState = TerminatedContractsUiState.Loading,
+    presenter = TerminatedContractsPresenter(getInsuranceContractsUseCaseProvider),
+  )
+
+internal class TerminatedContractsPresenter(
   private val getInsuranceContractsUseCaseProvider: Provider<GetInsuranceContractsUseCase>,
-) : ViewModel() {
-  private val retryChannel = RetryChannel()
+) : MoleculePresenter<TerminatedContractsEvent, TerminatedContractsUiState> {
+  @Composable
+  override fun MoleculePresenterScope<TerminatedContractsEvent>.present(
+    lastState: TerminatedContractsUiState,
+  ): TerminatedContractsUiState {
+    var dataLoadIteration by remember { mutableIntStateOf(0) }
+    var currentState by remember { mutableStateOf(lastState) }
 
-  val uiState: StateFlow<TerminatedContractsUiState> = flow {
-    emit(TerminatedContractsUiState.Loading)
-    getInsuranceContractsUseCaseProvider
-      .provide()
-      .invoke(forceNetworkFetch = false)
-      .onEach { result: Either<ErrorMessage, List<InsuranceContract>> ->
-        result.onLeft { errorMessage ->
-          logcat(LogPriority.INFO, errorMessage.throwable) {
-            "Terminated contracts failed to load ${errorMessage.message}"
-          }
-        }
+    CollectEvents { event ->
+      when (event) {
+        TerminatedContractsEvent.Retry -> dataLoadIteration++
       }
-      .map { insuranceContractResult ->
-        either {
-          val terminatedContracts = insuranceContractResult.bind().filter(InsuranceContract::isTerminated)
-          ensure(terminatedContracts.isNotEmpty()) {
-            ErrorMessage("", NoSuchElementException())
-          }
-          if (terminatedContracts.isEmpty()) {
-            logcat(LogPriority.ERROR) { "Terminated insurances screen got 0 terminated insurances" }
-            TerminatedContractsUiState.NoTerminatedInsurances
-          } else {
-            TerminatedContractsUiState.Success(terminatedContracts.toImmutableList())
-          }
-        }.fold(
-          ifLeft = { errorMessage ->
-            TerminatedContractsUiState.Error
-          },
-          ifRight = { it },
-        )
-      }
-      .collect(this)
-  }.stateIn(
-    viewModelScope,
-    SharingStarted.WhileSubscribed(5.seconds),
-    TerminatedContractsUiState.Loading,
-  )
+    }
 
-  fun retry() {
-    if (uiState.value is TerminatedContractsUiState.Error) {
-      retryChannel.retry()
+    LaunchedEffect(dataLoadIteration) {
+      if (lastState !is TerminatedContractsUiState.Success) {
```

Comment:
If the `lastState` was in fact success.
In the scenario where after the success, we still end up going into an error state, and then we do a retry after that, we will notice that the `lastState` which does not change was success, we will not see the state go from error to loading and then to the result of the use case. It would go from error to staying as error while we're loading, and then we'd get back the result, we wouldn't see the loading in-between.

I wrote this test
```kotlin
  @Test
  fun `with an initial success state, if there is an error, we are able to retry and get back in success state again`() = runTest {
    val getInsuranceContractsUseCase = FakeGetInsuranceContractsUseCase()
    val presenter = TerminatedContractsPresenter { getInsuranceContractsUseCase }
    presenter.test(TerminatedContractsUiState.Success(getInsuranceContractsUseCase.getTerminatedInsurances())) {
      assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Success>()
      getInsuranceContractsUseCase.addErrorToResponse()
      assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Error>()
      sendEvent(TerminatedContractsEvent.Retry)
      // assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Loading>()
      getInsuranceContractsUseCase.addTerminatedInsurancesToResponse()
      assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Success>()
    }
  }
```
And notice how it passes without there being a Loading state after the restart, despite us being in the error case and then doing restart. Specifically because of the bug I mention here. If you uncomment the line that checks for the error state the test will fail. Optimally it should pass. Perhaps you have to check on the `currentState` instead of `initialState` inside the LaunchedEffect.